ListUtil.xs

```diff
@@ -1357,75 +1348,98 @@ CODE:
+                 * The neatest approach I could find was provided by roboticus at:       *
+                 *     https://www.perlmonks.org/?node_id=11113490                       *
+                 * First, identify the lowest set bit and assign its value to an IV.     *
+                 * Note that this value will always be > 0, and always a power of 2.     */
+                IV lowest_set = iv & -iv;
+
+                /* Second, shift it left 53 bits to get location of the first bit        *
+                 * beyond arg's highest "allowed" set bit.                                                    *
+                 * NOTE: If lowest set bit is initially far enough left, then this left  *
+                 * shift operation will result in a value of 0, which is fine.           *
+                 * Then subtract 1 so that all of the ("allowed") bits below the set bit *
+                 * are 1 && all other ("disallowed") bits are set to 0.                  *
+                 * (If the value prior to subtraction was 0, then subtracting 1 will set *
+                 * all bits - which is also fine.)                                       */ 
+                UV valid_bits = (lowest_set << 53) - 1;
+
+                /* The value of arg can be exactly represented by a double unless one    *
+                 * or more of its "disallowed" bits are set - ie if iv & (~valid_bits)   *
+                 * is untrue. However, if (iv < 0 && !SvUOK(arg)) we need to multiply iv *
+                 * by -1 prior to performing that '&' operation - so multiply iv by sign.*/
+                if( !((iv * sign) & (~valid_bits)) ) {
+                    /* Avoid altering arg's flags */
+                    nv_arg = uok ? (NV)SvUV(arg) : (NV)SvIV(arg); 
                     sv_setpvn(keysv, (char *) &nv_arg, 8);
-                    sv_catpvn(keysv, "f", 1);
-                }
-                /* smaller floats get formatted using %g and could be equal to
-                 * a UV or IV */
+                }          
                 else {
-#if defined(MINGW_PERL_NO_ANSI)
-
-                   /* Because perl was not built with ansi compliance, doing: *
-                    * sv_setpvf(keysv, "%0.20" NVgf, nv_arg)                  *
-                    * will not always work as intended.                       *
-                    * But the following workaround does what we want.         */
-                    char buffer[32];
-                    sprintf(buffer, "%0.20" NVgf, nv_arg);
-                    sv_setpvf(keysv, "%s", buffer);
-#else
-                    sv_setpvf(keysv, "%0.20" NVgf, nv_arg);
-#endif
+                    /* Read in the bytes, rather than the numeric value of the IV/UV as  *
+                     * this is more efficient, despite having to sv_catpvn an extra byte.*/
+                    sv_setpvn(keysv, (char *) &iv, 8);
+                    /* We add an extra byte to distinguish between an IV/UV and an NV.   *
```

Comment:
Does this block suffer the problem that it considers all IVs distinct from all UVs? Does that mean it's possible to represent the same (positive) value as both an IV and a UV and yet they compare distinctly here?